gst/nnstreamer/tensor_crop/tensor_crop.c

```diff
@@ -187,14 +271,172 @@ gst_tensor_crop_get_property (GObject * object, guint prop_id,
+    GstEvent * event, gpointer user_data)
+{
+  GstTensorCropPadData *cpad;
+
+  g_return_val_if_fail (event != NULL, FALSE);
+
+  cpad = (GstTensorCropPadData *) data;
+
+  switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_CAPS:
+    {
+      GstCaps *caps;
+      GstStructure *structure;
+
+      gst_event_parse_caps (event, &caps);
+      structure = gst_caps_get_structure (caps, 0);
+
+      gst_tensors_config_from_structure (&cpad->config, structure);
+
+      gst_event_unref (event);
+      return gst_tensors_config_validate (&cpad->config);
+    }
+    default:
+      break;
+  }
+
+  return gst_collect_pads_event_default (pads, data, event, FALSE);
+}
+
+/**
+ * @brief Set pad caps if not negotiated.
  */
 static GstFlowReturn
-gst_tensor_crop_chain (GstPad * pad, GstObject * parent, GstBuffer * buffer)
+gst_tensor_crop_negotiate (GstTensorCrop * self)
+{
+  if (!gst_pad_has_current_caps (self->sinkpad_raw)) {
+    GST_ERROR_OBJECT (self,
+        "The raw pad of tensor_crop '%s' does not have pad caps.",
+        GST_ELEMENT_NAME (self));
+    return GST_FLOW_NOT_NEGOTIATED;
+  }
+
+  if (!gst_pad_has_current_caps (self->sinkpad_info)) {
+    GST_ERROR_OBJECT (self,
+        "The info pad of tensor_crop '%s' does not have pad caps.",
+        GST_ELEMENT_NAME (self));
+    return GST_FLOW_NOT_NEGOTIATED;
+  }
+
+  if (!gst_pad_has_current_caps (self->srcpad)) {
+    GstCaps *caps;
+    GstSegment segment;
+
+    if (self->send_stream_start) {
+      gchar *sid;
+
+      sid = g_strdup_printf ("%s-%08x",
+          GST_ELEMENT_NAME (self), g_random_int ());
+      gst_pad_push_event (self->srcpad, gst_event_new_stream_start (sid));
+      g_free (sid);
+
+      self->send_stream_start = FALSE;
+    }
+
+    /** @todo get config from collect-pads and set framerate */
+    caps = gst_caps_from_string (GST_TENSORS_FLEX_CAP_DEFAULT);
+    gst_caps_set_simple (caps, "framerate", GST_TYPE_FRACTION, 0, 1, NULL);
+
+    gst_pad_set_caps (self->srcpad, caps);
+    gst_caps_unref (caps);
+
+    gst_segment_init (&segment, GST_FORMAT_TIME);
+    gst_pad_push_event (self->srcpad, gst_event_new_segment (&segment));
+  }
+
+  return GST_FLOW_OK;
+}
+
+/**
+ * @brief Chain function called when the buffer is available on all of the collect pads.
+ */
+static GstFlowReturn
+gst_tensor_crop_collected (GstCollectPads * pads, gpointer user_data)
 {
   GstTensorCrop *self;
+  GstBuffer *raw_buffer, *info_buffer, *result;
+  GSList *walk;
+  GstFlowReturn ret;
+
+  self = GST_TENSOR_CROP (user_data);
+  raw_buffer = info_buffer = result = NULL;
+
+  ret = gst_tensor_crop_negotiate (self);
+  if (ret != GST_FLOW_OK)
+    return ret;
 
-  self = GST_TENSOR_CROP (parent);
+  for (walk = pads->data; walk; walk = g_slist_next (walk)) {
+    GstCollectData *data;
+
+    data = (GstCollectData *) walk->data;
+
+    if (data->pad == self->sinkpad_raw) {
+      raw_buffer = gst_collect_pads_pop (pads, data);
```

Comment:
break?

Reply:
Only two pads are added into collect-pad, so it is ok. plz see line 450.